kong/plugins/ai-proxy/drivers/anthropic.lua

```diff
@@ -0,0 +1,250 @@
+    prompt.max_tokens_to_sample = (model.options and model.options.max_tokens) or nil
+    prompt.model = model.name
+
+    return prompt, "application/json", nil
+  end,
+}
+
+local transformers_from = {
+  ["llm/v1/chat"] = function(response_string)
+    local response_table, err = cjson.decode(response_string)
+    if err then
+      return nil, "failed to decode cohere response"
+    end
+
+    if response_table.completion then
+      local res = {
+        choices = {
+          {
+            index = 0,
+            message = {
+              role = "assistant",
+              content = response_table.completion,
+            },
+            finish_reason = response_table.stop_reason,
+          },
+        },
+        model = response_table.model,
+        object = "chat.completion",
+      }
+        
+      return cjson.encode(res)
+    else
+      -- it's probably an error block, return generic error
+      return nil, "'completion' not in anthropic://llm/v1/chat response"
+    end
+  end,
+
+  ["llm/v1/completions"] = function(response_string)
+    local response_table, err = cjson.decode(response_string)
+    if err then
+      return nil, "failed to decode cohere response"
+    end
+
+    if response_table.completion then
+      local res = {
+        choices = {
+          {
+            index = 0,
+            text = response_table.completion,
+            finish_reason = response_table.stop_reason,
+          },
+        },
+        model = response_table.model,
+        object = "text_completion",
+      }
+
+      return cjson.encode(res)
+    else
+      -- it's probably an error block, return generic error
+      return nil, "'completion' not in anthropic://llm/v1/chat response"
+    end
+  end,
+}
+
+function _M.from_format(response_string, model_info, route_type)
+  -- MUST return a string, to set as the response body
+  ngx.log(ngx.DEBUG, "converting from ", model_info.provider, "://", route_type, " type to kong")
+
+  if not transformers_from[route_type] then
+    return nil, fmt("no transformer available from format %s://%s", model_info.provider, route_type)
+  end
+  
+  local ok, response_string, err = pcall(transformers_from[route_type], response_string)
+  if not ok or err then
+    return nil, fmt("transformation failed from type %s://%s: %s",
+                    model_info.provider,
+                    route_type,
+                    err or "unexpected_error"
+                )
+  end
+
+  return response_string, nil
+end
+
+function _M.to_format(request_table, model_info, route_type)
+  ngx.log(ngx.DEBUG, "converting from kong type to ", model_info.provider, "/", route_type)
+
+  if route_type == "preserve" then
+    -- do nothing
+    return request_table, nil, nil
+  end
+
+  if not transformers_to[route_type] then
+    return nil, nil, fmt("no transformer for %s://%s", model_info.provider, route_type)
+  end
+
+  local ok, request_object, content_type, err = pcall(
```

Comment:
Not necessarily a blocker, but I'm a little curious about the use of `pcall()` here and there throughout the drivers. Aren't we always calling our own code in these paths? And if so, can our own code not be written to be (reasonably) safe from throwing an exception? From a quick scan of most everything in `transformers_[from|to]`, it looks like it has solid error-handling (and you're wisely using `cjson.safe`).

Reply:
Just a total safety net for later, when I'm not working on it.